### ISSUE TITLE
Compilation fails on Ubuntu 18.04 with undefined reference to libldap with "external" libcurl

### ISSUE THREAD

**Issue by dopiera** [GoogleCloudPlatform/google-cloud-cpp#1714]
I spent some time trying to fix it, but I haven't figured it out yet - I'll need more time.

Workaround: supply `-DGOOGLE_CLOUD_CPP_CURL_PROVIDER=package` to cmake.

```
$ cat /etc/issue
Ubuntu 18.04.1 LTS \n \l
$ cmake -H. -Bbuild-output
$ cd build-output
$ make
[...]
../../../../external/lib/libcurl.a(ldap.c.o): In function `Curl_ldap':
ldap.c:(.text+0xa6): undefined reference to `ldap_url_parse'
ldap.c:(.text+0x112): undefined reference to `ldap_set_option'
ldap.c:(.text+0x121): undefined reference to `ldap_set_option'
ldap.c:(.text+0x12f): undefined reference to `ldap_init'
ldap.c:(.text+0x14d): undefined reference to `ldap_simple_bind_s'
ldap.c:(.text+0x169): undefined reference to `ldap_set_option'
ldap.c:(.text+0x177): undefined reference to `ldap_simple_bind_s'
ldap.c:(.text+0x1b2): undefined reference to `ldap_search_s'
ldap.c:(.text+0x1d3): undefined reference to `ldap_first_entry'
ldap.c:(.text+0x209): undefined reference to `ldap_get_dn'
ldap.c:(.text+0x281): undefined reference to `ldap_memfree'
ldap.c:(.text+0x294): undefined reference to `ldap_first_attribute'
ldap.c:(.text+0x2c6): undefined reference to `ldap_get_values_len'
ldap.c:(.text+0x47a): undefined reference to `ldap_set_option'
ldap.c:(.text+0x48b): undefined reference to `ldap_set_option'
ldap.c:(.text+0x496): undefined reference to `ldap_simple_bind_s'
ldap.c:(.text+0x4bd): undefined reference to `ldap_err2string'
ldap.c:(.text+0x4e8): undefined reference to `ldap_msgfree'
ldap.c:(.text+0x502): undefined reference to `ldap_free_urldesc'
ldap.c:(.text+0x513): undefined reference to `ldap_unbind_s'
ldap.c:(.text+0x58a): undefined reference to `ldap_value_free_len'
ldap.c:(.text+0x592): undefined reference to `ldap_memfree'
ldap.c:(.text+0x5a7): undefined reference to `ber_free'
ldap.c:(.text+0x5b4): undefined reference to `ldap_value_free_len'
ldap.c:(.text+0x5bc): undefined reference to `ldap_memfree'
ldap.c:(.text+0x601): undefined reference to `ldap_next_attribute'
ldap.c:(.text+0x61e): undefined reference to `ber_free'
ldap.c:(.text+0x62c): undefined reference to `ldap_next_entry'
ldap.c:(.text+0x67b): undefined reference to `ldap_err2string'
ldap.c:(.text+0x6bb): undefined reference to `ldap_err2string'
ldap.c:(.text+0x729): undefined reference to `ldap_memfree'
ldap.c:(.text+0x73f): undefined reference to `ldap_memfree'
ldap.c:(.text+0x779): undefined reference to `ldap_set_option'
ldap.c:(.text+0x788): undefined reference to `ldap_set_option'
ldap.c:(.text+0x7ac): undefined reference to `ldap_set_option'
ldap.c:(.text+0x7bd): undefined reference to `ldap_set_option'
ldap.c:(.text+0x7ca): undefined reference to `ldap_simple_bind_s'
collect2: error: ld returned 1 exit status
google/cloud/storage/examples/CMakeFiles/storage_quickstart.dir/build.make:102: recipe for target 'google/cloud/storage/examples/storage_quickstart' failed
make[2]: *** [google/cloud/storage/examples/storage_quickstart] Error 1
CMakeFiles/Makefile2:10782: recipe for target 'google/cloud/storage/examples/CMakeFiles/storage_quickstart.dir/all' failed
make[1]: *** [google/cloud/storage/examples/CMakeFiles/storage_quickstart.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

**Comment by coryan** [comment]
I just tested with:

```sh
cd ci/test-readme
sudo docker build -f Dockerfile.ubuntu .
```

and it compiled without problems.

I think libcurl automatically detects that libldap2 is installed, and compiles against it.  Our build scripts do not know that sometimes libcurl.a requires additional dependencies, and the build breaks.

I am going to test this idea and if it works I will disable the ldap functionality when building as an external package.

**Comment by remyabel** [comment]
libcurl has the strange behavior of having everything enabled by default and then checking if the dependency exists and disabling the flag if it doesn't. You have to explicitly pass `CURL_DISABLE_LDAP`. See https://github.com/curl/curl/blob/master/CMakeLists.txt#L432

```cmake
if(NOT CURL_DISABLE_LDAP)
# ...
    if(NOT HAVE_LDAP_H)
      message(STATUS "LDAP_H not found CURL_DISABLE_LDAP set ON")
      set(CURL_DISABLE_LDAP ON CACHE BOOL "" FORCE)
```

**Comment by coryan** [comment]
@remyabel yes.  In this case I think `HTTP_ONLY=ON` might be a more resilient fix, using external projects is great for development, and for testing, but I suspect most users will link against installed versions of libcurl when it comes time to use `google-cloud-cpp` in their projects. Curious about your thoughts on that actually.

**Comment by coryan** [comment]
Fixed by #1721.
